### DISCUSSION TITLE
Possibility to let users connect directly to DB

### DISCUSSION THREAD

**Discussion by bkniffler** [supabase/supabase#3525] [Questions] [answered]
Would be interesting, similar to stripes sigma (https://stripe.com/de/sigma), to allow users to directly connect to the database (e.g. for reporting). Is that possible with supabase built in integration, or would you consider to enable such a feature?

**Comment by silentworks** [comment, marked as answer]
Yes this is available by default, you can find the connection information under **settings/database** inside the Supabase Dashboard.

**Reply by burggraf** [reply]
When you say "directly connect to the database" this is the default, but you can connect through the client (i.e. the [Supabase Javscript Cient](https://supabase.io/docs/reference/javascript/select) or directly using a connection string.  The latter is pretty dangerous and not recommended, but most use cases are handled really well with the Supabase client.  If you need more complex SQL to be run, it's easy to write PostgreSQL functions (that can do literally anything) and call them with the client using the `.rpc()` method. [rpc](https://supabase.io/docs/reference/javascript/rpc)

**Reply by bkniffler** [reply]
Appreciate the response, I'm sorry, I think I didn't clearly ask the question @silentworks. I meant to let authenticated users (the ones with "authenticated" role, not admin users like myself) connect directly, while respecting their RLS policies. 

@burggraf What dangerous security implications would there be with such an approach, except for potentially exposing the DB to a lot of load due to no control over the queries people run?

**Reply by burggraf** [reply]
Giving users a direct channel to the database means you might open it up to users running arbitrary SQL commands, which creates a big security hole.  What exactly are you trying to accomplish?  There are probably simpler, more secure ways to do it.

**Reply by bkniffler** [reply]
Its clear to me that giving any users administrative access would cause a lot of trouble. But I thought that by using pg RLS features, we'd have a pretty solid isolation of data for users with the "authenticated" role. E.g., a user only being able to access todos with user_id=auth.id().

I'm thinking about introducing a feature like https://stripe.com/de/sigma for our end-users @burggraf.

**Reply by bkniffler** [reply]
I'm sure that the most sensible way to add such feature would be the db-per-user approach. But I was wondering if RLS would be a feasible way @burggraf.

**Reply by burggraf** [reply]
Is there any reason you can't implement what you're trying to do using the client library?  It's designed to work exactly as you're describing.  A user logs in, is authenticated, then you can use RLS to control access to all your tables.  Then then user can do exactly what they need to do and nothing else.  

Stripe integration is quite easy, and commonly used in Supabase projects.

**Reply by bkniffler** [reply]
I'm not talking about stripe in general, but about building something similar to stripe sigma. In my example, I'm stripe and I'd like to offer my clients a feature like sigma. I'm talking about adding a SQL editor that end users can use to build complex SQL queries from within our application.

But the more I think about this, the more I think it doesn't really make sense that way. I'll look into having a proxy endpoint that authenticated users can send queries into, that'll run the SQL commands within the users role and RLS scope.

**Reply by burggraf** [reply]
`proxy endpoint that authenticated users can send queries into`
That makes a lot more sense, I think, security-wise.

**Reply by steve-chavez** [reply]
> I'll look into having a proxy endpoint that authenticated users can send queries into, that'll run the SQL commands within the users role and RLS scope.

@bkniffler This could be a function that executes dynamic sql based on the text arg you pass to it. If you do it like that, it would already be executed with the authenticated role and apply RLS.

> What dangerous security implications would there be with such an approach, except for potentially exposing the DB to a lot of load due to no control over the queries people run?

Of course, that would be the major issue here since users can potentially run expensive operations/infinite loops - which should be mitigated with extra safeguards.
